### PR TITLE
Create E2E package

### LINKED ISSUE #6807: Create E2E test package
This is a parent issue for the following tasks

- [ ] Create a npm package `@woocommerce/admin-e2e-tests`
- [ ] Move the existing code to the package and update `specs` to use the package

### PR DESCRIPTION
**Pull request by rrennick**
See #6807

This PR moves the E2E tests and supporting code to `@woocommerce/admin-e2e-tests` located in `packages/admin-e2e-tests`. There are no functional changes to any of the tests. The changes include

- Exclude `admin-e2e-tests` from the build zip
- Update the `test:e2e` script
  - The test container expects `WP_VERSION` to be an explicit version and defaults to `latest`
  - Add a check for running containers before attempting to stop them
- Addresses all linting issues in the E2E code
- Adds `admin-e2e-tests` to `npm run build`

I set the initial version to `0.1.0` with the intent to release a subsequent version with the remainder of the tasks from #6807.

Note: `jest` prohibits `import`ing its tests functions outside of the jest environment. Any files using jest have Common JS references for those functions.
 
### Detailed test instructions:

-   Review CI run
-   Locally run `npm run test:e2e`

No changelog

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -60,7 +60,7 @@
 		"pretest": "npm run -s install-if-no-packages",
 		"test": "./node_modules/jest-24.9.0/bin/jest.js --config tests/js/jest.config.json",
 		"test:debug": "node --inspect-brk ./node_modules/jest-24.9.0/bin/jest.js --config tests/js/jest.config.json --watch --runInBand --no-cache",
-		"test:e2e": "WP_VERSION=latest npm run build && npx wc-e2e docker:down && npx wc-e2e docker:up && npx wc-e2e test:e2e",
+		"test:e2e": "npm run build && [[$(docker ps | grep woocommerce-admin) == '']] || npx wc-e2e docker:down && npx wc-e2e docker:up && npx wc-e2e test:e2e",
```
Comment (adrianduffell) [inline package.json:63]: Could a more unique string than `woocommerce-admin` be used? I think this will also match other Docker services we run.

Comment (rrennick) [reply package.json:63]: > Could a more unique string than woocommerce-admin be used?

Yes, we can change it with an `appName` entry in `tests/e2e/config/default.json`. Would `woocommerce-admin-e2e` work?

Comment (rrennick) [reply package.json]: @adrianduffell I updated the PR to change the container prefix to `woocommerce-admin-e2e`. The small catch that goes with the change is that people will need to stop their existing e2e containers with `docker-compose`.

Comment (adrianduffell) [reply package.json]: @rrennick It doesn't seem ideal to use `docker-compose` to stop the containers, but I don't think it is a blocker for this PR.

What's the reason for it? Could it be improved in the future?

File: packages/admin-e2e-tests/package.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ "license": "GPL-3.0+",
+ "main": "build/index.js",
+ "types": "build/index.d.ts",
+ "files": [
+  "/build/",
+  "!*.ts.map",
+  "!*.tsbuildinfo"
+ ],
+ "sideEffects": false,
+ "scripts": {
```
Comment (adrianduffell) [inline packages/admin-e2e-tests/package.json:24]: Are these scripts needed? I don't think the other packages have their own scripts.

File: packages/admin-e2e-tests/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (adrianduffell) [inline packages/admin-e2e-tests/tsconfig.json:1]: Is this file needed? I think just using the main `tsconfig.json` might be preferable to keep all packages on the same config. @samueljseay what do you think?

**Review by adrianduffell — CHANGES_REQUESTED** [review]
Thanks @rrennick, this tested well locally 👍 

It doesn't look like the CI checks have run yet. I think you will need to merge the latest `main` changes to get those to work.

I left some suggestions mostly about TypeScript.

File: packages/admin-e2e-tests/tsconfig.json (hunk 3 above)
Comment (samueljseay) [reply packages/admin-e2e-tests/tsconfig.json:1]: I think this is ok @adrianduffell every package will have its own basic settings that varies slightly from the others and it opens up the option to disable or enable some rules if we have to on a per package basis. What I'm glad to see here is that it extends the base config. :+1:

File: package.json (hunk 1 above)
Comment (rrennick) [reply package.json]: > What's the reason for it?

It's a one time thing. The current containers (if someone has them running) are prefixed with `woocommerce-admin`. The script to stop them will be looking for `woocommerce-admin-e2e`. Since the old prefix doesn't match, the e2e `docker:down` won't attempt to stop them. The new ones won't be able to start because the port is already allocated.

Comment (adrianduffell) [reply package.json]: Ah I see, thanks,I didn't understand it would be a one time thing. That sounds perfectly fine!

File: packages/admin-e2e-tests/src/elements/DropdownTypeaheadField.ts [hunk 4]
```diff
@@ -1,6 +1,12 @@
-import { clearAndFillInput } from '@woocommerce/e2e-utils';
+/**
+ * Internal dependencies
+ */
 import { BaseElement } from './BaseElement';
 
+/* eslint-disable @typescript-eslint/no-var-requires */
```
[the remark this answers is not in the document]
Comment (rrennick) [reply packages/admin-e2e-tests/src/elements/DropdownTypeaheadField.ts:6]: Jest doesn't permit that: https://github.com/facebook/jest/blob/master/packages/jest-globals/src/index.ts#L29:L31

File: tests/e2e/specs/activate-and-setup/basic-setup.test.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+const { testAdminBasicSetup } = require( '@woocommerce/admin-e2e-tests' );
```
[the remark this answers is not in the document]
Comment (rrennick) [reply tests/e2e/specs/activate-and-setup/basic-setup.test.js:1]: > Could the files in tests/e2e/specs/ be kept as TypeScript?

@adrianduffell Unfortunately no. If you pull the latest of this branch and create another TS test file where you import and call one of the test functions then `npm run build`, the build will fail. That's due to not being able to build with jest functions outside of the jest environment.

File: packages/admin-e2e-tests/package.json (hunk 2 above)
Comment (rrennick) [reply packages/admin-e2e-tests/package.json]: > Are these scripts needed?

@adrianduffell I removed these and it looks fine. This is ready for another review. Note that I updated some of the package manifests to eliminate linting warnings. The linting log was over 1000 lines and the missing `@wordpress/*` package dependencies were contributing a considerable amount.

File: tests/e2e/specs/activate-and-setup/basic-setup.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rrennick) [reply tests/e2e/specs/activate-and-setup/basic-setup.test.js:1]: Also, if I need to add any changelog entries please let me know.

File: packages/admin-e2e-tests/package.json (hunk 2 above)
Comment (adrianduffell) [reply packages/admin-e2e-tests/package.json]: @rrennick Thanks! I'll review this later today.

Comment (adrianduffell) [reply packages/admin-e2e-tests/package.json]: Thanks for fixing the linting log 👍

File: tests/e2e/specs/activate-and-setup/basic-setup.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrianduffell) [reply tests/e2e/specs/activate-and-setup/basic-setup.test.js:1]: I don't think we'll need more changelog entries as the WC Admin product doesn't change. The changelog entry for the new package looks sufficient.

**Comment by adrianduffell** [comment]
@rrennick You can override the `Lint the changelog` check by adding the text "No changelog" to a new line in the PR description.

File: tests/e2e/specs/activate-and-setup/basic-setup.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrianduffell) [reply tests/e2e/specs/activate-and-setup/basic-setup.test.js:1]: > If you pull the latest of this branch and create another TS test file where you import and call one of the test functions then npm run build, the build will fail.

I wasn't able to reproduce this. I renamed this file to `basic-setup.test.tsx` and then `npm run build` ran successfully. I was also able to create a new file. What error do you see?

File: packages/admin-e2e-tests/src/elements/DropdownTypeaheadField.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianduffell) [reply packages/admin-e2e-tests/src/elements/DropdownTypeaheadField.ts:6]: Ah ok, no worries.

**Review by adrianduffell — COMMENTED** [review]
@rrennick Nice job on this! Thanks for addressing the feedback, I think it's pretty close. I still wonder if we're able to keep the files in `tests/e2e/specs` as TypeScript because I couldn't reproduce the build error.

We have a goal to eventually migrate everything to TypeScript so ideally they'd remain like that, but on the other hand I don't think it's worth holding up this PR over it.

File: tests/e2e/specs/activate-and-setup/basic-setup.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rrennick) [reply tests/e2e/specs/activate-and-setup/basic-setup.test.js]: This is the error I was getting locally after adding the new test added in June - https://github.com/woocommerce/woocommerce-admin/pull/7084/checks?check_run_id=3022700048#step:5:1029

**Comment by rrennick** [comment]
@adrianduffell Thanks for all the feedback and suggestions on this. It's ready for another review.

**Review by adrianduffell — APPROVED** [review]
@rrennick  Awesome, the latest changes look good to me!

Note: A fix for the extension count test is being worked on by the team as well. We're making the tests more flexible now that the onboarding wizard is populated from a remote source and can change at any time.

Thanks for your work on this! 🚀
